### PR TITLE
[REVIEW] Add Doxygen configuration file and build Doxygen HTML

### PR DESCRIPTION
**Pull request by jrhemstad**
Closes https://github.com/rapidsai/cudf/issues/698

In an attempt to start building the HTML version of our incomplete Doxygen source code documentation, this PR does the following:

- [x] Adds Doxygen configuration file `cudf/cpp/doxygen/Doxyfile`

- [x] Setup configuration file to build HTML from Doxygen source code comments

- [x] Add `make dox` Cmake target

~ Explore what other configuration options are necessary for optimal HTML documentation ~

I have no experience with building Doxygen HTML, so I could use an assist from someone who perhaps has more experience or feels like diving in and learning. Open issues:

- [x] Current HTML output includes Arrow Doxygen documentation. Need to figure out how to exclude that.

Ultimately, building this documentation should be made part of the CI process (ping @raydouglass @mike-wendt ). Doxygen also throws warnings for any undocumented code, which could be used in the future to disallow merging code that hasn't been properly documented. 

**Update:**

After adding the Cmake `make doc` target, I have no further plans to add anything additional to this PR. There is plenty of remaining work required in improving the Doxygen documentation present in the libcudf source to improve what the built HTML looks like, but that is orthogonal to this PR.

### CODE REVIEW

**Comment by j-ieong** [comment]
Looks cool! Comments:
- No graphs for class hierarchy, struct collaboration, etc. in yours. I installed graphviz to get them
- You can use `EXCLUDE_xxx` to not include those files and/or namespaces. Setting `EXCLUDE_SYMBOLS = org::apache` works for me

**Comment by harrism** [comment]
Shouldn't you put the output html directory in .gitignore and not commit those files?

**Comment by jrhemstad** [comment]
> Shouldn't you put the output html directory in .gitignore and not commit those files?

Indeed, done.

File: cpp/doxygen/Doxyfile [hunk 1]
```diff
@@ -0,0 +1,2427 @@
...
+# The EXCLUDE_SYMBOLS tag can be used to specify one or more symbol names
+# (namespaces, classes, functions, etc.) that should be excluded from the
+# output. The symbol name can be a fully qualified name, a word, or if the
+# wildcard * is used, a substring. Examples: ANamespace, AClass,
+# AClass::ANamespace, ANamespace::*Test
+#
+# Note that the wildcards are matched against the file with absolute path, so to
+# exclude all test directories use the pattern */test/*
+
+EXCLUDE_SYMBOLS        =
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/doxygen/Doxyfile]: Awesome, thanks.

File: cpp/doxygen/Doxyfile [hunk 2]
```diff
@@ -0,0 +1,2427 @@
...
+# The default value is: NO.
+
+INLINE_INHERITED_MEMB  = NO
+
+# If the FULL_PATH_NAMES tag is set to YES, doxygen will prepend the full path
+# before files name in the file list and in the header files. If set to NO the
+# shortest path that makes the file name unique will be used
+# The default value is: YES.
+
+FULL_PATH_NAMES        = YES
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/doxygen/Doxyfile]: Good idea, I think removing the full path names is good.

**Comment by harrism** [comment]
I think this just needs a `make doc` target in CMakeLists.txt, and then we can merge assuming we are happy with the appearance of the docs.

**Pull request merged**